src/client/datascience/kernel-launcher/localKernelFinder.ts

```diff
@@ -0,0 +1,654 @@
+        try {
+            // Get list of all of the specs
+            const kernels = await this.listKernels(resource, cancelToken);
+
+            // Always include the interpreter in the search if we can
+            const interpreter =
+                option && isInterpreter(option)
+                    ? option
+                    : resource && this.extensionChecker.isPythonExtensionInstalled
+                    ? await this.interpreterService.getActiveInterpreter(resource)
+                    : undefined;
+
+            // Find the preferred kernel index from the list.
+            const notebookMetadata = option && !isInterpreter(option) ? option : undefined;
+            const preferred = findPreferredKernel(kernels, resource, [], notebookMetadata, interpreter, undefined);
+            if (preferred) {
+                traceInfoIf(
+                    !!process.env.VSC_JUPYTER_LOG_KERNEL_OUTPUT,
+                    `findKernel found ${getDisplayNameOrNameOfKernelConnection(preferred)}`
+                );
+                return preferred as LocalKernelConnectionMetadata;
+            }
+        } catch (e) {
+            traceError(`findKernel crashed: ${e} ${e.stack}`);
+            return undefined;
+        }
+    }
+
+    // Search all our local file system locations for installed kernel specs and return them
+    @captureTelemetry(Telemetry.KernelListingPerf)
+    public async listKernels(
+        resource: Resource,
+        cancelToken?: CancellationToken
+    ): Promise<LocalKernelConnectionMetadata[]> {
+        try {
+            // Get an id for the workspace folder, if we don't have one, use the fsPath of the resource
+            const workspaceFolderId =
+                this.workspaceService.getWorkspaceFolderIdentifier(
+                    resource,
+                    resource?.fsPath || this.workspaceService.rootPath
+                ) || 'root';
+
+            // If we have not already searched for this resource, then generate the search
+            if (workspaceFolderId && !this.workspaceToMetadata.has(workspaceFolderId)) {
+                this.workspaceToMetadata.set(
+                    workspaceFolderId,
+                    this.findResourceKernelMetadata(resource, cancelToken).then((items) => {
+                        traceInfoIf(
+                            !!process.env.VSC_JUPYTER_LOG_KERNEL_OUTPUT,
+                            `Kernel specs for ${resource?.toString() || 'undefined'} are \n ${JSON.stringify(
+                                items,
+                                undefined,
+                                4
+                            )}`
+                        );
+                        return items;
+                    })
+                );
+            }
+
+            this.writeCache().ignoreErrors();
+
+            // ! as the has and set above verify that we have a return here
+            return await this.workspaceToMetadata.get(workspaceFolderId)!;
```

Comment:
Why do we need to do an await here?

Reply:
Otherwise the catch doesn't work. The promise is not waited on here, but rather on the caller.

Reply:
Thanks for explaining 😊

Reply:
This is an easy pattern to miss. (I've had this problem during this refactor in fact).

If you have something like so:

```typescript
try {
    return functionThatReturnsAPromise();
} catch (e) {
   trace(`Swallowing e: ${e}`);
}
```

The catch will never be hit. That's because the promise is returned. Without the await the potential rejection is never thrown in this function but rather in the first spot that awaits the promise.